Behaviors/SoftDeleteBehavior.php

```diff
@@ -0,0 +1,211 @@
+        
+        // Do not do anything, if the base object of the data sheet is not the object with the behavior and is not
+        // extended from it.
+        if (! $eventData->getMetaObject()->isExactly($this->getObject())) {
+            return;
+        }
+
+        // prevent deletion of the main object, but dont prevent the cascading deletion
+        $event->preventDelete(false);
+
+        $transaction = $event->getTransaction();
+
+        $affected_rows = 0;
+
+        $updateData = $eventData->copy();
+
+        // remove all columns, except of system columns (like 'id' or 'modified-on' columns)
+        foreach ($updateData->getColumns() as $col){
+            if ($updateData->getMetaObject()->hasAttribute($col->getAttributeAlias())){
+                switch (true){
+                    case ($col->getAttribute()->isSystem()):
+                        break;
+                    default:
+                        $updateData->getColumns()->remove($col);
+                }
+            } else {
+                $updateData->getColumns()->remove($col);
+            }
+        }
+
+        // first check if metaobject has soft-delete-attribute
+        if ($eventData->getMetaObject()->hasAttribute($this->getSoftDeleteAttributeAlias())){
+            
+            // add the soft-delete-column
+            $deletedCol = $updateData->getColumns()->addFromAttribute($this->getSoftDeleteAttribute());
+            
+            // if there are no datarows in the passed datasheet, but there are filters assigned:
+            // add a single row of data, with only the soft-delete-attribute being set, so that this 
+            // attribute can be assigned to every row fitting the filter later
+            if ($updateData->isEmpty() && $updateData->getFilters() !== null){
+                $updateData->addRow([$this->getSoftDeleteAttributeAlias() => $this->getSoftDeleteValue()]);
+            }
+            
+            // if the datasheet still contains no datarows, then no items have to be marked as deleted
+            if ($updateData->isEmpty() === false){
+                $deletedCol->setValueOnAllRows($this->getSoftDeleteValue());
+                $updatedRows = $updateData->dataUpdate(false, $transaction);
+                $affected_rows += $updatedRows;
+            }
+        }
+            
+        $eventData->setCounterForRowsInDataSource($updateData->countRowsInDataSource());
+
+        // also update the original data sheet for further use
+        if ($eventData->isEmpty() === false){
+            $deletedCol = $eventData->getColumns()->getByAttribute($this->getSoftDeleteAttribute());
+            $deletedCol->setValueOnAllRows($this->getSoftDeleteValue());
+        }
+
+        return $affected_rows;
+    }
+    
+    /**
+     * 
+     * @return string
+     */
+    public function getSoftDeleteAttributeAlias() 
+    {
+        if ($this->getObject()->hasAttribute($this->soft_delete_attribute_alias)){
```

Comment:
Ich würde diesen Check in getSoftDeleteAttribute(). Dort kommt ja in jedem Fall eine exception die musst du bloß in deine neu exception einpacken. So wie jetzt wird der Check doppelt ausgeführt. Das ist unnötig.

Mache die Methode auch protected.